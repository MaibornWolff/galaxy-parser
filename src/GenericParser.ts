import { McCabeComplexity } from "./metrics/McCabeComplexity";
import { Functions } from "./metrics/Functions";
import { Classes } from "./metrics/Classes";
import { grammars } from "./grammars";
import { getParseFile } from "./helper";
import path from "path";
import fs from "fs";
import {ExpressionMetricMapping} from "./app";
import {LinesOfCode} from "./metrics/LinesOfCode";
import {CommentLines} from "./metrics/CommentLines";
import {RealLinesOfCode} from "./metrics/RealLinesOfCode";
import {Coupling} from "./metrics/Coupling";
import {CouplingCSharp} from "./metrics/CouplingCSharp";

export class GenericParser {
    private readonly fileMetrics: Metric[] = [];
    private readonly comprisingMetrics: CouplingMetric[] = [];

    private edgeMetrics: CouplingMetricResult;

    constructor() {
        const nodeTypesJson = fs.readFileSync(fs.realpathSync("./resources/node-types-mapped.config")).toString();
        const allNodeTypes: ExpressionMetricMapping[] = JSON.parse(nodeTypesJson);

        this.fileMetrics = [
            new McCabeComplexity(allNodeTypes),
            new Functions(allNodeTypes),
            new Classes(allNodeTypes),
            new LinesOfCode(allNodeTypes),
            new CommentLines(allNodeTypes),
            new RealLinesOfCode(allNodeTypes),
        ];

        this.comprisingMetrics = [
            new CouplingCSharp(allNodeTypes)
        ];
    }

    getEdgeMetrics(): CouplingMetricResult {
        return this.edgeMetrics
    }

    calculateMetrics(sourcesRoot: string) {
        sourcesRoot = fs.realpathSync(sourcesRoot);

        console.log("\n\n");
        console.log("----- Try to parse " + sourcesRoot + " recursively -----");
        console.log("\n\n");

        const startTime = performance.now();

        const files = this.findFilesRecursively(
            sourcesRoot,
            [...grammars.keys()],
            ["node_modules", ".idea", "dist", "build", "out", "vendor"],
            []
        );

        console.log(" --- " + files.length + " files detected");
        console.log("\n\n");

        const fileMetrics = new Map<string, Map<string, MetricResult>>();

        const parseFiles = [];
        for (const filePath of files) {
            const parseFile = getParseFile(filePath);
            if (!parseFile || !grammars.has(parseFile.language)) {
                continue;
            }

            parseFiles.push(parseFile);
        }

        for (const parseFile of parseFiles) {
            const metricResults = new Map<string, MetricResult>();
            fileMetrics.set(parseFile.filePath, metricResults)

            console.log(
                " ------------ Parsing File " +
                    path.basename(parseFile.filePath) +
                    "  ------------ "
            );

            for (const metric of this.fileMetrics) {
                const metricResult = metric.calculate(parseFile);
                metricResults.set(metricResult.metricName, metricResult);
            }
        }

        for (const metric of this.comprisingMetrics) {
            this.edgeMetrics = metric.calculate(parseFiles);
        }

        const endTime = performance.now();
        const duration = endTime - startTime;

        console.log("\n\n");
        console.log(
            `Parsing took ${duration / 1000} seconds or ${
                duration / 1000 / 60
            } minutes respectively`
        );

        return fileMetrics;
    }

    private findFilesRecursively(
        dir,
        supportedFileExtensions: string[] = [],
        excludedFolders: string[] = [],
        fileList: string[] = []
    ) {
        if (fs.lstatSync(dir).isFile()) {
            return [dir];
        }

        fs.readdirSync(dir).forEach((file) => {
            if (fs.statSync(path.join(dir, file)).isDirectory()) {
                const isPathExcluded = excludedFolders.some((folder) => {
                    return dir.includes(folder);
                });
                if (isPathExcluded) {
                    return;
                }

                fileList = this.findFilesRecursively(
                    path.join(dir, file),
                    supportedFileExtensions,
                    excludedFolders,
                    fileList
                );
            } else if (supportedFileExtensions.includes(file.split(".").pop().toLowerCase())) {
                fileList = fileList.concat(path.join(dir, file));
            }
        });
        return fileList;
    }
}
