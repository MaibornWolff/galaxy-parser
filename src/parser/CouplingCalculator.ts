import fs from "fs";
import { ExpressionMetricMapping } from "./helper/Model";
import { Configuration } from "./Configuration";
import { Coupling } from "./metrics/coupling/Coupling";
import { NamespaceCollector } from "./resolver/NamespaceCollector";
import { UsagesCollector } from "./resolver/UsagesCollector";
import { CouplingMetric, ParseFile } from "./metrics/Metric";
import { PublicAccessorCollector } from "./resolver/PublicAccessorCollector";
import { debuglog } from "node:util";

const dlog = debuglog("metric-gardener");

export class CouplingCalculator {
    private readonly comprisingMetrics: CouplingMetric[] = [];
    private config: Configuration;

    private readonly namespaceCollector: NamespaceCollector;
    private readonly publicAccessorCollector: PublicAccessorCollector;
    private readonly usageCollector: UsagesCollector;

    constructor(configuration: Configuration) {
        this.config = configuration;

        const nodeTypesJson = fs
            .readFileSync(fs.realpathSync("./resources/node-types-mapped.config"))
            .toString();
        const allNodeTypes: ExpressionMetricMapping[] = JSON.parse(nodeTypesJson);

        this.namespaceCollector = new NamespaceCollector();
        this.publicAccessorCollector = new PublicAccessorCollector();
        this.usageCollector = new UsagesCollector();

        this.comprisingMetrics = [
            new Coupling(
                allNodeTypes,
                this.namespaceCollector,
                this.usageCollector,
                this.publicAccessorCollector
            ),
        ];
    }

    calculateMetrics(parseFiles: ParseFile[]) {
        const sourcesRoot = fs.realpathSync(this.config.sourcesPath);

        dlog("\n\n");
        dlog("----- Parsing Coupling of files in " + sourcesRoot + " recursively -----");
        dlog("\n\n");

        dlog(" --- " + parseFiles.length + " files detected", "\n\n");

        // TODO rewrite this to support multiple coupling metrics
        return this.comprisingMetrics[0].calculate(parseFiles);
    }
}
