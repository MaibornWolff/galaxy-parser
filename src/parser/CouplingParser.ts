import fs from "fs";
import { ExpressionMetricMapping } from "./helper/Model";
import { Configuration } from "./Configuration";
import { Coupling } from "./metrics/coupling/Coupling";
import neo4j from "neo4j-driver";
import { NamespaceCollector } from "./collectors/NamespaceCollector";
import { UsagesCollector } from "./collectors/UsagesCollector";
import { CouplingMetric, CouplingMetricValue, ParseFile } from "./metrics/Metric";

export class CouplingParser {
    private readonly comprisingMetrics: CouplingMetric[] = [];
    private config: Configuration;

    private readonly namespaceCollector: NamespaceCollector;
    private readonly usageCollector: UsagesCollector;

    constructor(configuration: Configuration) {
        this.config = configuration;

        const nodeTypesJson = fs
            .readFileSync(fs.realpathSync("./resources/node-types-mapped.config"))
            .toString();
        const allNodeTypes: ExpressionMetricMapping[] = JSON.parse(nodeTypesJson);

        this.namespaceCollector = new NamespaceCollector();
        this.usageCollector = new UsagesCollector();

        this.comprisingMetrics = [
            new Coupling(allNodeTypes, this.namespaceCollector, this.usageCollector),
        ];
    }

    calculateMetrics(parseFiles: ParseFile[]) {
        const sourcesRoot = fs.realpathSync(this.config.sourcesPath);

        console.log("\n\n");
        console.log("----- Parsing Coupling of files in " + sourcesRoot + " recursively -----");
        console.log("\n\n");

        console.log(" --- " + parseFiles.length + " files detected", "\n\n");

        // TODO rewrite this to support multiple coupling metrics
        const couplingMetrics = this.comprisingMetrics[0].calculate(parseFiles);

        if (this.config.persistDependencyGraph) {
            this.buildDependencyGraph(couplingMetrics.relationships).then(() => {
                console.log(
                    "Dependency Graph done: Visit http://localhost:7474 with login admin/admin"
                );
            });
        }

        return couplingMetrics;
    }

    private async buildDependencyGraph(relationships: CouplingMetricValue[]) {
        // create dependency graph in neo4j
        // write nodes and edges to neo4j
        if (!this.config.persistDependencyGraph) {
            return;
        }

        const driver = neo4j.driver("neo4j://localhost:7687", neo4j.auth.basic("neo4j", "admin"));
        const session = driver.session();

        try {
            await session.writeTransaction((tx) => tx.run("MATCH (n) DETACH DELETE n"));

            for (const [
                language,
                namespaceReferences,
            ] of this.namespaceCollector.getAllNamespaces()) {
                for (const namespaceReferencesPerFile of namespaceReferences.values()) {
                    for (const namespaceReference of namespaceReferencesPerFile.values()) {
                        const { namespace, source, className, classType, namespaceDelimiter } =
                            namespaceReference;

                        await session.writeTransaction((tx) =>
                            tx.run(
                                `
                                    CREATE (n:` +
                                    classType.toUpperCase() +
                                    `)
                                    SET
                                        n.namespace = $namespace,
                                        n.sourcePath = $source,
                                        n.className = $className,
                                        n.language = $language,
                                        n.fullyQualifiedName = $fullyQualifiedName
                                `,
                                {
                                    namespace,
                                    source,
                                    className,
                                    language,
                                    fullyQualifiedName: namespace + namespaceDelimiter + className,
                                }
                            )
                        );
                    }
                }
            }

            for (const relationship of relationships) {
                const { fromNamespace, toNamespace, usageType } = relationship;

                await session.writeTransaction((tx) =>
                    tx.run(
                        `
                            MATCH
                                (a),
                                (b)
                            WHERE a.fullyQualifiedName = $fromNamespace AND b.fullyQualifiedName = $toNamespace
                            CREATE (a)-[r:` +
                            usageType.toUpperCase() +
                            `]->(b)
                        `,
                        { fromNamespace, toNamespace, usageType }
                    )
                );
            }
        } finally {
            await session.close();
        }

        // on application exit:
        await driver.close();
    }
}
