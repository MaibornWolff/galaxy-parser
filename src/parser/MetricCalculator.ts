import { McCabeComplexity } from "./metrics/McCabeComplexity";
import { Functions } from "./metrics/Functions";
import { Classes } from "./metrics/Classes";
import fs from "fs";
import { LinesOfCode } from "./metrics/LinesOfCode";
import { CommentLines } from "./metrics/CommentLines";
import { RealLinesOfCode } from "./metrics/RealLinesOfCode";
import { ExpressionMetricMapping } from "./helper/Model";
import { Configuration } from "./Configuration";
import { Metric, MetricResult, ParseFile } from "./metrics/Metric";
import nodeTypesConfig from "./config/nodeTypesConfig.json";

/**
 * Arranges the calculation of all basic single file metrics on multiple files.
 */
export class MetricCalculator {
    private readonly fileMetrics: Metric[] = [];
    private config: Configuration;

    /**
     * Constructs a new instance of {@link MetricCalculator} for arranging the calculation of all
     * basic single file metrics on multiple files.
     * @param configuration Configuration that should be applied for this new instance.
     */
    constructor(configuration: Configuration) {
        this.config = configuration;
        const allNodeTypes: ExpressionMetricMapping[] =
            nodeTypesConfig as ExpressionMetricMapping[];

        this.fileMetrics = [
            new McCabeComplexity(allNodeTypes),
            new Functions(allNodeTypes),
            new Classes(allNodeTypes),
            new LinesOfCode(),
            new CommentLines(allNodeTypes),
            new RealLinesOfCode(allNodeTypes),
        ];
    }

    /**
     * Calculates all basic single file metrics on the specified files.
     * @param parseFiles Files for which the metrics should be calculated.
     * @return Map that relates the file path of each file to a map of the calculated metrics.
     */
    calculateMetrics(parseFiles: ParseFile[]) {
        const sourcesRoot = fs.realpathSync(this.config.sourcesPath);

        console.log(
            "\n\n",
            "----- Calculating metrics for " + sourcesRoot + " recursively -----",
            "\n\n"
        );
        console.log(" --- " + parseFiles.length + " files detected", "\n\n");

        const fileMetrics = new Map<string, Map<string, MetricResult>>();

        for (const parseFile of parseFiles) {
            const metricResults = new Map<string, MetricResult>();
            fileMetrics.set(parseFile.filePath, metricResults);

            console.log(" ------------ Parsing File " + parseFile.filePath + ":  ------------ ");

            for (const metric of this.fileMetrics) {
                const metricResult = metric.calculate(parseFile);
                metricResults.set(metricResult.metricName, metricResult);
            }
        }

        return fileMetrics;
    }
}
