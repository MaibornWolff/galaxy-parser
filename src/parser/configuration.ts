/**
 * Parameters of the constructor of {@link Configuration}.
 * Represents configuration options that can be provided by the user via command line arguments.
 */
export type ConfigurationParameters = {
    /**
     * Path to the source files to be parsed.
     */
    sourcesPath: string;
    /**
     * Path where the output file with the calculated metrics should be stored.
     */
    outputPath: string;
    /**
     * Whether dependencies should be analyzed.
     */
    parseDependencies: boolean;
    /**
     * Folders to exclude from being searched for files to be parsed.
     */
    exclusions: string;
    /**
     * Whether to parse all .h files as C instead of C++.
     */
    parseAllHAsC: boolean;
    /**
     * Folders or file names for which .h files should be parsed as C instead of C++ files. Ignored if parseAllAsC is set.
     */
    parseSomeHAsC: string;
    /**
     * Whether to compress the output file with the calculated metrics in a zip archive.
     */
    compress: boolean;
    /**
     * Whether to include the relative file paths or absolute paths of the analyzed files in the output.
     */
    relativePaths: boolean;
};

/**
 * Configures the files to be parsed and the metrics to be calculated.
 */
export class Configuration {
    /**
     * Resolved, absolute path to the source files to be parsed.
     */
    readonly sourcesPath: string;

    /**
     * Path where the output file with the calculated metrics should be stored.
     */
    readonly outputPath: string;

    /**
     * Whether dependencies should be analyzed.
     */
    readonly parseDependencies: boolean;

    /**
     * Folders to exclude from being searched for files to be parsed.
     */
    readonly exclusions: Set<string>;

    /**
     * Whether to parse all .h files as C instead of C++.
     */
    readonly parseAllHAsC: boolean;

    /**
     * Folders or file names for which .h files should be parsed as C instead of C++ files.
     */
    readonly parseSomeHAsC: Set<string>;

    /**
     * Whether to compress the output file with the calculated metrics in a zip archive.
     */
    readonly compress: boolean;

    /**
     * Whether to include the relative file paths or absolute paths of the analyzed files in the output.
     */
    readonly relativePaths: boolean;

    /**
     * Constructs a new {@link Configuration} object by specifying the configuration options passed by the user
     * as command line arguments.
     * @param parameters {@link Parameters} object containing the configuration options.
     */
    constructor(parameters: ConfigurationParameters) {
        this.sourcesPath = parameters.sourcesPath;
        this.outputPath = parameters.outputPath;
        this.parseDependencies = parameters.parseDependencies;

        this.exclusions = new Set(
            parameters.exclusions.length > 0
                ? parameters.exclusions.split(",").map((exclusion) => exclusion.trim())
                : null,
        );

        this.parseAllHAsC = parameters.parseAllHAsC;

        this.parseSomeHAsC = new Set(
            !parameters.parseAllHAsC && parameters.parseSomeHAsC.length > 0
                ? parameters.parseSomeHAsC
                      .split(",")
                      .map((folderOrFileName) => folderOrFileName.trim())
                : null,
        );

        this.compress = parameters.compress;
        this.relativePaths = parameters.relativePaths;
    }
}
