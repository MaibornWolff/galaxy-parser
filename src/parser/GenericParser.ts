import { grammars } from "./helper/Grammars";
import { findFilesRecursively } from "./helper/Helper";
import fs from "fs";
import { Configuration } from "./Configuration";
import { MetricCalculator } from "./MetricCalculator";
import { CouplingCalculator } from "./CouplingCalculator";
import { CouplingResult, MetricResult } from "./metrics/Metric";

export class GenericParser {
    private readonly config: Configuration;

    constructor(configuration: Configuration) {
        this.config = configuration;
    }

    calculateMetrics() {
        const startTime = performance.now();

        const parseFiles = findFilesRecursively(
            fs.realpathSync(this.config.sourcesPath),
            [...grammars.keys()],
            this.config.exclusions,
            []
        );

        console.log(" --- " + parseFiles.length + " files detected", "\n\n");

        let fileMetrics = new Map<string, Map<string, MetricResult>>();
        if (this.config.parseMetrics) {
            const metricsParser = new MetricCalculator(this.config);
            fileMetrics = metricsParser.calculateMetrics(parseFiles);
        }

        let couplingMetrics = {} as CouplingResult;
        if (this.config.parseDependencies) {
            const couplingParser = new CouplingCalculator(this.config);
            couplingMetrics = couplingParser.calculateMetrics(parseFiles);
        }

        console.log("Final Coupling Metrics", couplingMetrics);

        const endTime = performance.now();
        const duration = endTime - startTime;

        console.log("\n\n");
        console.log("#####################################");
        console.log("#####################################");
        console.log("\n\n");
        console.log(
            `Parsing took ${duration / 1000} seconds or ${
                duration / 1000 / 60
            } minutes respectively`
        );

        return {
            fileMetrics,
            couplingMetrics,
        };
    }
}
