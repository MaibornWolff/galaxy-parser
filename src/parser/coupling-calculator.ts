import { type Configuration } from "./configuration.js";
import { Coupling } from "./metrics/coupling/coupling.js";
import { TypeCollector } from "./resolver/type-collector.js";
import { UsagesCollector } from "./resolver/usages-collector.js";
import {
    type CouplingMetric,
    type CouplingResult,
    ParsedFile,
    type SourceFile,
} from "./metrics/metric.js";
import { PublicAccessorCollector } from "./resolver/public-accessor-collector.js";

export class CouplingCalculator {
    private readonly comprisingMetrics: CouplingMetric[] = [];
    private readonly config: Configuration;

    private readonly typeCollector: TypeCollector;
    private readonly publicAccessorCollector: PublicAccessorCollector;
    private readonly usageCollector: UsagesCollector;

    constructor(configuration: Configuration) {
        this.config = configuration;
        this.typeCollector = new TypeCollector();
        this.publicAccessorCollector = new PublicAccessorCollector();
        this.usageCollector = new UsagesCollector();
        this.comprisingMetrics = [
            new Coupling(
                this.config,
                this.typeCollector,
                this.usageCollector,
                this.publicAccessorCollector,
            ),
        ];
    }

    processFile(sourceFile: SourceFile): void {
        if (this.config.parseDependencies && sourceFile instanceof ParsedFile) {
            this.comprisingMetrics[0].processFile(sourceFile);
        }
    }

    calculateMetrics(): CouplingResult {
        if (this.config.parseDependencies) {
            console.log("Calculating coupling metrics...");
            // TODO rewrite this to support multiple coupling metrics
            return this.comprisingMetrics[0].calculate();
        }

        return { relationships: [], metrics: new Map() };
    }
}
