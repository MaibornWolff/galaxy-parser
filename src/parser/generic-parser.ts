import process from "node:process";
import pMap from "p-map";
import { findFilesAsync, formatPrintPath } from "./helper/helper.js";
import { type Configuration } from "./configuration.js";
import { calculateMetrics } from "./metric-calculator.js";
import { CouplingCalculator } from "./coupling-calculator.js";
import {
    type SourceFile,
    type FileMetricResults,
    ErrorFile,
    UnsupportedFile,
    type CouplingResult,
} from "./metrics/metric.js";
import { parse } from "./helper/tree-parser.js";

/**
 * Arranges the parsing of files and calculation of metrics as specified by the stored configuration.
 */
export class GenericParser {
    /**
     * Configuration according to which files are parsed and metrics are calculated.
     * @private
     */
    private readonly config: Configuration;

    /**
     * Constructs a new {@link GenericParser} object with the supplied configuration applied
     * (see also {@link Configuration}).
     * @param configuration The configuration to apply for the new {@link GenericParser} object.
     */
    constructor(configuration: Configuration) {
        this.config = configuration;
    }

    /**
     * Parses files and calculates metrics as specified by the configuration of this {@link GenericParser} object.
     */
    async calculateMetrics(): Promise<{
        couplingMetrics: CouplingResult;
        fileMetrics: Map<string, FileMetricResults>;
        unsupportedFiles: string[];
        errorFiles: string[];
    }> {
        const filePaths = await this.loadFilePaths();

        const couplingParser = new CouplingCalculator(this.config);

        let parsed = 0;
        const results = await pMap(
            filePaths,
            async (filePath) => {
                const sourceFile = await parse(filePath, this.config);
                couplingParser.processFile(sourceFile);

                const progress = Math.floor((parsed++ / filePaths.length) * 100);
                showProgressBar(progress);

                return calculateMetrics(sourceFile);
            },
            { concurrency: 10 },
        );
        clearProgressBar();

        const couplingMetrics = couplingParser.calculateMetrics();
        return { ...this.processResults(results), couplingMetrics };
    }

    private async loadFilePaths(): Promise<string[]> {
        const filePaths: string[] = [];

        for await (const filePath of findFilesAsync(this.config)) {
            filePaths.push(filePath);
            if (filePaths.length % 1000 === 0) {
                process.stdout.write(`\rfiles: ${filePaths.length.toString()}`);
            }
        }

        console.log(`\rfiles: ${filePaths.length.toString()}`);

        return filePaths;
    }

    private processResults(results: Array<[SourceFile, FileMetricResults]>): {
        fileMetrics: Map<string, FileMetricResults>;
        unsupportedFiles: string[];
        errorFiles: string[];
    } {
        const fileMetrics = new Map<string, FileMetricResults>();
        const unsupportedFiles: string[] = [];
        const errorFiles: string[] = [];

        for (const [sourceFile, fileMetricResults] of results) {
            const printPath = formatPrintPath(sourceFile.filePath, this.config);

            if (sourceFile instanceof ErrorFile) {
                errorFiles.push(printPath);
                console.error("Error while parsing the syntax tree for the file " + printPath);
                console.error(sourceFile.error);
            } else {
                fileMetrics.set(printPath, fileMetricResults);

                // Inform about errors that occurred while calculating (some) metrics on the syntax tree
                for (const metricError of fileMetricResults.metricErrors) {
                    console.error(
                        "Error while calculating the metric " +
                            metricError.metricName +
                            " on the file " +
                            sourceFile.filePath,
                    );
                    console.error(metricError.error);
                }

                if (sourceFile instanceof UnsupportedFile) {
                    unsupportedFiles.push(printPath);
                }
            }
        }

        return {
            fileMetrics,
            unsupportedFiles,
            errorFiles,
        };
    }
}

let progress = 0;
function showProgressBar(i: number): void {
    if (i > progress) {
        progress = i;
        const dots = ".".repeat(i);
        const left = 100 - i;
        const empty = " ".repeat(left);
        process.stdout.write(`\r[${dots}${empty}] ${i.toString()}%`);
    }
}

function clearProgressBar(): void {
    process.stdout.write("\r" + " ".repeat(110) + "\r");
    progress = 0;
}
