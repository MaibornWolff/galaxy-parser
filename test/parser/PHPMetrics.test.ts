import { getParserConfiguration, sortCouplingResults, testFileMetrics } from "./TestHelper";
import fs from "fs";
import { GenericParser } from "../../src/parser/GenericParser";

describe("PHP metrics tests", () => {
    const phpTestResourcesPath = "./resources/php/";

    describe("parses PHP McCabeComplexity metric", () => {
        it("should count if statements correctly", async () => {
            await testFileMetrics(phpTestResourcesPath + "if-statements.php", "mcc", 7);
        });

        it("should count functions and methods correctly", async () => {
            await testFileMetrics(phpTestResourcesPath + "functions-and-methods.php", "mcc", 7);
        });

        it("should not count multiple return statements within functions and methods like sonar", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "multiple-return-statements.php",
                "mcc",
                3
            );
        });

        it("should not count any class declaration", async () => {
            await testFileMetrics(phpTestResourcesPath + "classes.php", "mcc", 0);
        });

        it("should count case statements correctly", async () => {
            await testFileMetrics(phpTestResourcesPath + "case-statements.php", "mcc", 3);
        });

        it("should count try-catch-finally properly", async () => {
            await testFileMetrics(phpTestResourcesPath + "throw-try-catch-finally.php", "mcc", 2);
        });

        it("should count loops properly", async () => {
            await testFileMetrics(phpTestResourcesPath + "loops.php", "mcc", 4);
        });
    });

    describe("parses PHP classes metric", () => {
        it("should count class declarations", async () => {
            await testFileMetrics(phpTestResourcesPath + "classes.php", "classes", 3);
        });
    });

    describe("parses PHP functions metric", () => {
        it("should count function declarations", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "functions-and-methods.php",
                "functions",
                7
            );
        });
    });

    describe("parses PHP lines of code metric", () => {
        it("should count number of lines correctly for a non-empty file with empty last line", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "empty-last-line.php",
                "lines_of_code",
                66
            );
        });

        it("should count number of lines correctly for a non-empty file with non-empty last line", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "php-example-code.php",
                "lines_of_code",
                65
            );
        });

        it("should count number of lines correctly for an empty file", async () => {
            await testFileMetrics(phpTestResourcesPath + "empty.php", "lines_of_code", 1);
        });

        it("should count number of lines correctly for an file with one non-empty line", async () => {
            await testFileMetrics(phpTestResourcesPath + "one-line.php", "lines_of_code", 1);
        });

        it("should count number of lines correctly for an file with just a line break", async () => {
            await testFileMetrics(phpTestResourcesPath + "line-break.php", "lines_of_code", 2);
        });
    });

    describe("parses PHP real lines of code metric", () => {
        it("should count correctly for a non-empty file, ignoring comments and empty lines", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "php-example-code.php",
                "real_lines_of_code",
                43
            );
        });

        it("should count correctly if there is a comment in the same line as actual code", async () => {
            await testFileMetrics(
                phpTestResourcesPath + "same-line-comment.php",
                "real_lines_of_code",
                11
            );
        });
    });

    describe("parses PHP commentLines metric", () => {
        it(
            "should count number of comment lines correctly, including line with curly brackets and comment " +
                "lines inside block comment",
            async () => {
                await testFileMetrics(
                    phpTestResourcesPath + "php-example-code.php",
                    "comment_lines",
                    12
                );
            }
        );
    });

    describe("parsing PHP dependencies", () => {
        it("should calculate the right dependencies and coupling metrics", async () => {
            const realInputPath = fs.realpathSync(phpTestResourcesPath + "coupling-examples/");
            const parser = new GenericParser(getParserConfiguration(realInputPath, true, true));

            const results = await parser.calculateMetrics();
            const couplingResult = results.couplingMetrics;
            sortCouplingResults(couplingResult);

            expect(couplingResult).toMatchSnapshot();
        });
    });
});
