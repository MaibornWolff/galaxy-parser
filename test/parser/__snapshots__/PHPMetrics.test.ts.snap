// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PHP metrics tests parsing PHP dependencies should calculate the right dependencies and coupling metrics 1`] = `
Object {
  "metrics": Map {
    "AnotherController.php" => Object {
      "coupling_between_objects": 4,
      "incoming_dependencies": 0,
      "instability": 1,
      "outgoing_dependencies": 4,
    },
    "AnotherControllerInterface.php" => Object {
      "coupling_between_objects": 1,
      "incoming_dependencies": 1,
      "instability": 0,
      "outgoing_dependencies": 0,
    },
    "BlubController.php" => Object {
      "coupling_between_objects": 5,
      "incoming_dependencies": 2,
      "instability": 0.6,
      "outgoing_dependencies": 3,
    },
    "ControllerInterface.php" => Object {
      "coupling_between_objects": 3,
      "incoming_dependencies": 3,
      "instability": 0,
      "outgoing_dependencies": 0,
    },
    "FastControllerInterface.php" => Object {
      "coupling_between_objects": 3,
      "incoming_dependencies": 1,
      "instability": 0.6666666666666666,
      "outgoing_dependencies": 2,
    },
    "Library\\\\Helper.php" => Object {
      "coupling_between_objects": 4,
      "incoming_dependencies": 3,
      "instability": 0.25,
      "outgoing_dependencies": 1,
    },
    "Library\\\\HelperOutput.php" => Object {
      "coupling_between_objects": 2,
      "incoming_dependencies": 1,
      "instability": 0.5,
      "outgoing_dependencies": 1,
    },
  },
  "relationships": Array [
    Object {
      "fromClassName": "AnotherControllerInterface",
      "fromNamespace": "App\\\\FastControllerInterface",
      "fromSource": "FastControllerInterface.php",
      "toClassName": "AnotherControllerInterface",
      "toNamespace": "App\\\\AnotherControllerInterface",
      "toSource": "AnotherControllerInterface.php",
      "usageType": "implements",
    },
    Object {
      "fromClassName": "ControllerInterface",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\AnotherControllerOne",
      "fromSource": "AnotherController.php",
      "toClassName": "ControllerInterface",
      "toNamespace": "App\\\\ControllerInterface",
      "toSource": "ControllerInterface.php",
      "usageType": "implements",
    },
    Object {
      "fromClassName": "ControllerInterface",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\BlubControllerOne1",
      "fromSource": "BlubController.php",
      "toClassName": "ControllerInterface",
      "toNamespace": "App\\\\ControllerInterface",
      "toSource": "ControllerInterface.php",
      "usageType": "implements",
    },
    Object {
      "fromClassName": "ControllerInterface",
      "fromNamespace": "App\\\\FastControllerInterface",
      "fromSource": "FastControllerInterface.php",
      "toClassName": "ControllerInterface",
      "toNamespace": "App\\\\ControllerInterface",
      "toSource": "ControllerInterface.php",
      "usageType": "implements",
    },
    Object {
      "fromClassName": "BlubControllerOne1",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\AnotherControllerOne",
      "fromSource": "AnotherController.php",
      "toClassName": "BlubControllerOne1",
      "toNamespace": "App\\\\CouplingExamplesOne\\\\BlubControllerOne1",
      "toSource": "BlubController.php",
      "usageType": "extends",
    },
    Object {
      "fromClassName": "BlubControllerOne1",
      "fromNamespace": "App\\\\CouplingExamplesTwo\\\\BlubControllerTwo1",
      "fromSource": "BlubController.php",
      "toClassName": "BlubControllerOne1",
      "toNamespace": "App\\\\CouplingExamplesOne\\\\BlubControllerOne1",
      "toSource": "BlubController.php",
      "usageType": "extends",
    },
    Object {
      "fromClassName": "Helper",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\AnotherControllerOne",
      "fromSource": "AnotherController.php",
      "toClassName": "Helper",
      "toNamespace": "App\\\\CouplingExamples\\\\Library\\\\Helper",
      "toSource": "Library\\\\Helper.php",
      "usageType": "usage",
    },
    Object {
      "fromClassName": "Helper",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\BlubControllerOne1",
      "fromSource": "BlubController.php",
      "toClassName": "Helper",
      "toNamespace": "App\\\\CouplingExamples\\\\Library\\\\Helper",
      "toSource": "Library\\\\Helper.php",
      "usageType": "usage",
    },
    Object {
      "fromClassName": "Helper",
      "fromNamespace": "App\\\\CouplingExamples\\\\Library\\\\HelperOutput",
      "fromSource": "Library\\\\HelperOutput.php",
      "toClassName": "Helper",
      "toNamespace": "App\\\\CouplingExamples\\\\Library\\\\Helper",
      "toSource": "Library\\\\Helper.php",
      "usageType": "usage",
    },
    Object {
      "fromClassName": "HelperOutput",
      "fromNamespace": "App\\\\CouplingExamples\\\\Library\\\\Helper",
      "fromSource": "Library\\\\Helper.php",
      "toClassName": "HelperOutput",
      "toNamespace": "App\\\\CouplingExamples\\\\Library\\\\HelperOutput",
      "toSource": "Library\\\\HelperOutput.php",
      "usageType": "usage",
    },
    Object {
      "fromClassName": "FastControllerInterface",
      "fromNamespace": "App\\\\CouplingExamplesOne\\\\AnotherControllerOne",
      "fromSource": "AnotherController.php",
      "toClassName": "FastControllerInterface",
      "toNamespace": "App\\\\FastControllerInterface",
      "toSource": "FastControllerInterface.php",
      "usageType": "implements",
    },
  ],
}
`;
